From 987217414baba61cf188cf96e94026dc9ce78f81 Mon Sep 17 00:00:00 2001
From: "hanghang.luo" <hanghang.luo@amlogic.com>
Date: Thu, 13 Jul 2023 02:23:06 +0000
Subject: [PATCH 1/4] gst-plugin-aml-v4l2dec: CB1 calculate the fence num and
 send event to sink [1/1]

PD#SWPL-129180

Problem:
What the fence default value should be set to?

Solution:
The value of the fence is obtained by the decoder by sending the event

Verify:
AM301

Change-Id: I099a24fcd3df3cd4c3689225a79a94aed21ae17b
Signed-off-by: hanghang.luo <hanghang.luo@amlogic.com>
---
 src/gstamlv4l2videodec.c | 33 +++++++++++++++++++++++++++------
 1 file changed, 27 insertions(+), 6 deletions(-)

diff --git a/src/gstamlv4l2videodec.c b/src/gstamlv4l2videodec.c
index df05674..99c36f6 100644
--- a/src/gstamlv4l2videodec.c
+++ b/src/gstamlv4l2videodec.c
@@ -817,6 +817,24 @@ gst_v4l2_drop_event (GstAmlV4l2Object * v4l2object)
   return;
 }
 
+static void
+gst_aml_v4l2_video_dec_set_fence(GstVideoDecoder *decoder,GstVideoInfo *info)
+{
+    GstAmlV4l2VideoDec *self = GST_AML_V4L2_VIDEO_DEC(decoder);
+    GstStructure *s;
+    GstEvent *event;
+
+    guint fence_num = (info->interlace_mode == GST_VIDEO_INTERLACE_MODE_INTERLEAVED) ?
+        GST_AML_V4L2_DEFAULT_CAP_BUF_MARGIN : self->v4l2capture->min_buffers-2;
+    s = gst_structure_new ("video_fence","fence_num",G_TYPE_UINT,fence_num,NULL);
+    if (s)
+    {
+        event = gst_event_new_custom (GST_EVENT_CUSTOM_DOWNSTREAM, s);
+        GST_DEBUG_OBJECT(self,"Send video_fence Event: %"GST_PTR_FORMAT,event);
+        gst_pad_push_event (decoder->srcpad, event);
+    }
+}
+
 static void
 gst_aml_v4l2_video_dec_loop(GstVideoDecoder *decoder)
 {
@@ -867,11 +885,14 @@ gst_aml_v4l2_video_dec_loop(GstVideoDecoder *decoder)
             peer = gst_pad_get_peer (decoder->srcpad);
             if (peer)
             {
-              s = gst_structure_new_empty ("IS_SVP");
-              event = gst_event_new_custom (GST_EVENT_CUSTOM_DOWNSTREAM, s);
-              gst_pad_send_event (peer, event);
-              GST_DEBUG_OBJECT(self, "Send SVP Event");
-              gst_object_unref (peer);
+                s = gst_structure_new_empty ("IS_SVP");
+                if (s)
+                {
+                    event = gst_event_new_custom (GST_EVENT_CUSTOM_DOWNSTREAM, s);
+                    gst_pad_send_event (peer, event);
+                    GST_DEBUG_OBJECT(self, "Send SVP Event");
+                }
+                gst_object_unref (peer);
             }
         }
 
@@ -928,7 +949,7 @@ gst_aml_v4l2_video_dec_loop(GstVideoDecoder *decoder)
         else
             gst_aml_v4l2_clear_error(&error);
         gst_caps_unref(caps);
-
+        gst_aml_v4l2_video_dec_set_fence(decoder,&info);
         output_state = gst_video_decoder_set_output_state(decoder,
                                                           info.finfo->format, info.width, info.height, self->input_state);
 
-- 
2.25.1

